### DISCUSSION TITLE
Field Array: Sort entries with Array.sort

### DISCUSSION THREAD

**Discussion by knoefel** [react-hook-form/react-hook-form#4264] [Q&A] [answered]
First of all, and even though it has been said many times before, thank you for this great library!!

I have a Field Array, which i want to sort by date. The items have for e.g. the following simplified shape:

```typescript
interface item {
  name: string;
  date: string;
}
```

I initialise the form with `initialvalues` prop from the `useForm` hook, so that my Field Array is filled with entries. This data is not sorted yet (it would also not be the solution to sort the array before the initial call of `useForm`, as it must also be possible to add new field array entries at runtime, and these would then also have to be sorted again). After the initialising i sort the Field Array entries and then loop over them to build the form list:

```typescript
const sortedFields = [...fields].sort((a, b) => a.date.localeCompare(b.date));

<ul>
  {sortedFields.map((item, index) => 
    (
      <li key={item.id}>
        <input
          name={`test[${index}].name`}
          defaultValue={`${item.name}`}
          ref={register()}
        />
        <input
          name={`test[${index}].date`}
          defaultValue={`${item.date}`}
          ref={register()}
         />
      </li>  
    )
  )}
</ul>
```
The problem is, when i change for example the name field in any Field Array item, the change is sometimes made to the wrong Field Array Item. I guess that index is the problem, as i change it with the `Array.sort` method, so that internal state from `useForm` can't assign the change to the right item.

So my question is: how can i order the Field Array and guarantee it at the same time that the right item form fields are updated? 

Or to put it more generally: can I use `Array.sort` for this or should I go another way?

I've seen the `swap` and `move` methods, but felt it would be even more cumbersome this way? Or would this be the right way?

Thanks in advance for feedback and input!

**Comment by knoefel** [comment, marked as answer]
Ok, i solved it by saving each index and `keyName` of the unsorted `fields` in a `Map`:

```typescript
const indexes = new Map(fields.map(({ id }, index) => [id, index]));

// Alternative as an object
const indexes = fields.reduce((map, { id }, index) => ({...map, [id]: index}) , {});
```

And then in the loop i can access the right index:

```typescript
// Map Version
name={`test[${indexes.get(item.id)}].name`}

// object Version
name={`test[${indexes[item.id]}].name`}
```

But since I don't find the method very elegant and it feels a bit hacky, I would really appreciate some feedback on how maybe the 'right' way to deal with sorting array fields would be. Many thanks in advance.

**Reply by bluebill1049** [reply]
we don't have `sort` out the box, what you did above seems reasonable.

**Reply by stuarthallows** [reply]
Would it be possible to go the _controlled field array_ approach in order to custom sort an array?
https://react-hook-form.com/api/usefieldarray

**Reply by dnlrsr** [reply]
@bluebill1049 is there a way to request this as a feature?

**Reply by Moshyfawn** [reply]
See the #7785 feature request, @Dinodanio
